src/servers/src/proto.rs

```diff
@@ -36,46 +37,88 @@ pub struct PromLabel {
 }
 
 impl Clear for PromLabel {
-    fn clear(&mut self) {
-        self.name.clear();
-        self.value.clear();
-    }
+    fn clear(&mut self) {}
 }
 
 impl PromLabel {
-    pub fn merge_field<B>(
+    pub fn merge_field(
         &mut self,
         tag: u32,
         wire_type: WireType,
-        buf: &mut B,
-        ctx: DecodeContext,
-    ) -> Result<(), DecodeError>
-    where
-        B: Buf,
-    {
+        buf: &mut Bytes,
+    ) -> Result<(), DecodeError> {
         const STRUCT_NAME: &str = "PromLabel";
         match tag {
             1u32 => {
                 // decode label name
                 let value = &mut self.name;
-                prost::encoding::bytes::merge(wire_type, value, buf, ctx).map_err(|mut error| {
+                merge_bytes(value, buf).map_err(|mut error| {
                     error.push(STRUCT_NAME, "name");
                     error
                 })
             }
             2u32 => {
                 // decode label value
                 let value = &mut self.value;
-                prost::encoding::bytes::merge(wire_type, value, buf, ctx).map_err(|mut error| {
+                merge_bytes(value, buf).map_err(|mut error| {
                     error.push(STRUCT_NAME, "value");
                     error
                 })
             }
-            _ => prost::encoding::skip_field(wire_type, tag, buf, ctx),
+            _ => prost::encoding::skip_field(wire_type, tag, buf, Default::default()),
         }
     }
 }
 
+#[inline(always)]
+fn copy_to_bytes(data: &mut Bytes, len: usize) -> Bytes {
```

Comment:
This function seems to have different behavior depending on `len`. The returned `Bytes` might be either an owned or weak reference. Maybe also add some doc to it?

Reply:
> This function seems to have different behavior depending on `len`. The returned `Bytes` might be either an owned or weak reference. Maybe also add some doc to it?

Either case is safe as long as we guarantee that returned bytes does not outlive original bytes.